django/template/defaultfilters.py

```diff
@@ -184,6 +184,7 @@ def floatformat(text, arg=-1):
     units = len(tupl[1])
     units += -tupl[2] if m else tupl[2]
     prec = abs(p) + units + 1
+    prec = max(1, prec)  # Prevent out-of-range precision.
```

Comment:
One more idea :smile: Since `1` is the lowest value, wouldn't it be better to use `decimal.getcontext().prec` instead?

Reply:
`Context` was added to fix ticket-15789.

Reply:
For the max value? So, `prec = max(1, decimal.getcontext().prec`? On my system precision is set to 28 by default, and I noticed that the testsuite fails when using that default value.

While investigating this bug, I had experimented with removing the entire calculation and just using the default value for `quantize`. 

`floatformat(-1.323297138040798e35, 2)` fails when doing that.

Reply:
I was thinking about `prec = max(decimal.getcontext().prec, prec)`

Reply:
Ah, I see. To allow one to set a  minimal precision?

I've adapted the code accordingly.

Reply:
For what it's worth, I've also compared using the maximum precision (thus skipping the entire calculation of the needed precision), and it  seems to have minimal performance impact (as it is measured in ns).

```python
In [22]: %timeit Decimal("99999.999").quantize(Decimal('0.1'), context=Context(prec=10))
504 ns ± 6.96 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [23] %timeit Decimal("99999.999").quantize(Decimal('0.1'), context=Context(prec=decimal.MAX_PREC))
519 ns ± 0.234 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

I've also tried measure an approximation of the code, to get the performance impact of the latest change:

```python
In [40]: %timeit max(getcontext().prec, abs(2)+len(Decimal("99999.99").as_tuple()[1])+4 if True else 2+1)
457 ns ± 0.126 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [41]: %timeit max(1, abs(2)+len(Decimal("99999.99").as_tuple()[1])+4 if True else 2+1)
422 ns ± 0.11 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Reply:
I read parts of the implementation of Decimal in CPython in `Lib/_pydecimal.py` and noticed this method:
```python
def adjusted(self):
        """Return the adjusted exponent of self"""
        try:
            return self._exp + len(self._int) - 1
        # If NaN or Infinity, self._exp is string
        except TypeError:
            return 0
```

Instead of the current calculation of `prec`, we could do:

```python
if d == 0:
    prec = p
else:
    prec = max(1, d.normalize().adjusted() + 2 + abs(p))
```

Or similarly:
```python
prec = p
if d != 0:
    prec = max(1, d.normalize().adjusted() + 2 + abs(p))
```

Performance seems much improved and the tests succeed. I haven't measured the impact on actual template
rendering when templates contain lots of `floatformat` filters.

```python
In [42]: %timeit max(1, Decimal("99999.999").normalize().adjusted()+2+abs(2))
238 ns ± 0.14 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

But, frankly, I'm not sure if its worth the trouble... Maybe just using MAX_PREC is easier, and maybe even faster as the entire calculation can then be removed from the code.

Reply:
@takis Thanks :+1: Let's merge and backport the current version, and we can discuss potential improvements/simplifications in a separate ticket.

Can I ask you to create a follow up ticket and a summary of your investigation?

Reply:
Hi @felixxm,

Great!

Sure, I'll create a follow up ticket.